text/0039-reduce-module-size.md

```diff
@@ -0,0 +1,518 @@
+| bundled npm dependencies       | 0.69%                          |
+| .ts-fixture                    | 0.37%                          |
+| .jsiirc                        | 0.36%                          |
+
+
+
+### Why should we _not_ do this?
+
+The main reason to not do this is that it will inevitably break up the
+monolothic `aws-cdk-lib` package in some manner, and we will lose some of the
+wonderful simplicity of releasing and using a single package. But, we can't keep
+everything in one place, and reduce the size enough to have a significant
+positive impact.
+
+### What is the technical solution (design) of this feature?
+
+There are different categories of files contributing to the size of aws-cdk-lib,
+and each category will have it’s own solution. The following sections cover the
+top five categories in size contribution in order from largest to smallest.
+
+#### .jsii Files
+
+There are two .jsii files bundled in `aws-cdk-lib` for the JSII runtime to work
+for any non-NodeJS languages. And, to enable code and documentation generation
+from the published npm package.
+
+|                 | Size (MB) | Percentage of aws-cdk-lib |
+| --------------- | --------- | ------------------------- |
+| .jsii.tabl.json | 55.1      | 22.84 %                   |
+| .jsii           | 45.35     | 18.8 %                    |
+
+We will compress the .jsii.tabl.json and .jsii files. Compressing with gzip
+currently creates a 4 MB and 4.7 MB file, respectively. This change will impact
+the JSII runtime, and Construct Hub. NodeJS JSII runtimes don't need to use
+either file, and non-NodeJS runtimes would decompress the .jsii file when
+needed. Construct Hub generates documentation from the .jsii.tabl.json file, and
+it will also decompress this file as needed.
+
+#### Source Map Files (.js.map)
+
+These files currently makeup 22% of the total size of `aws-cdk-lib`. Source map
+files map from the transformed source to the original source. In the case of
+`aws-cdk-lib`, these are not actually useful today, since we do not ship the
+original .ts files. We will remove the source maps from the released package.
+
+#### Lambda Layer Zip Files
+
+There are three zip files included in the `aws-cdk-lib` package. They are listed in the table below.
+
+| Filename                                    | Includes                                                                             | Usage                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | Size (MB) | Percentage of aws-cdk-lib |
```

Comment:
minor nit: these calculations (along with the .jsii files one) suggests that the full module size is calculated as ~241 MB, not ~238 MB. I think we should replace the places that say 238 with 241.

Reply:
The current total size according to npm is 238 MB, so I'd like to leave the total at that. So, there is some other discrepancy in my numbers, I'm not sure if it's worth it to fix that. 

These calculations are approximate. The goal of having numbers here at all is to give the reader an idea for which solutions are highest priority, and whether they will actually succeed in reducing the package size significantly. The precision of the numbers is not as important. Maybe we should include a disclaimer here that all of the numbers are a little bit rough.

Reply:
For me, first thing I did when I read 

> Note: The percentages do not add up to 100. Some files, e.g. various .json
> files, are excluded and contribute very little to the size.

was add up all the numbers to see exactly what percentage of the total size was excluded. Maybe that's just me, but I figure I'd add in a "total" section in that table in case other people have the same thought as me.

Reply:
Another reason for discrepancies in numbers is I calculated most of them a few releases ago, and the package size changes with every release.

Reply:
Added an Appendix that hopefully make this all WAY more clear. Please add more comments if not.